### ISSUE TITLE
fix: Run 8 loss divergence after step 220 — lr=2e-5 too aggressive for 22K dataset

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#274]
## Problem

Run 8 (Qwen3-4B NF4 QLoRA, conversations_v4.jsonl) diverges after step 220:
- Steps 1-220: loss 15.50 → 1.43 (excellent convergence)
- Steps 220-416: loss 1.43 → 10.03 (diverging toward random baseline 11.93)
- Gradient norms rising: 7 → 15 → 70 → 100+
- Embed gradient norms exploding: 1.2M → 318M

Run killed at step 416. Checkpoint-200 (loss ~1.43) is the best artifact.

## Loss Trajectory

```
step=1    loss=15.50  gnorm=42.7
step=55   loss=13.00  gnorm=14.0
step=110  loss=3.80   gnorm=24.4
step=165  loss=2.39   gnorm=7.1
step=220  loss=1.43   gnorm=7.7    ← BEST
step=275  loss=5.82   gnorm=15.6   ← diverging
step=330  loss=7.05   gnorm=15.3
step=385  loss=10.03  gnorm=70.7   ← near random baseline
```

## Root Cause Analysis

1. **lr=2e-5 too high** after warmup completes at step 100. The model overfits on easy examples quickly (loss drops to 1.43 by step 220), then gradient norms explode on harder examples.

2. **No validation-based early stopping** in current implementation — the patience=3 config is not wired into the CudaTransformerTrainer.

3. **22K dataset heterogeneity** — the dataset mixes trivial safe examples (variable assignments) with complex unsafe patterns (injection, race conditions). After learning the easy patterns, the model destabilizes on hard ones.

## Proposed Fix

1. **Lower lr to 5e-6** (4x reduction) — matches successful LoRA fine-tuning literature for 4B models
2. **Implement cosine annealing** — lr should decay after warmup, not stay constant
3. **Wire validation early stopping** — evaluate on val split every 200 steps, stop when val loss increases 3x
4. **Resume from checkpoint-200** — don't retrain from scratch, warm-start from best checkpoint

## Config for Run 9

```yaml
optimizer:
  lr: 0.000005          # 5e-6 (was 2e-5)
scheduler:
  name: cosine           # decay after warmup
  warmup:
    steps: 50            # shorter warmup (was 100)
training:
  epochs: 1              # single pass (was 3)
  resume: checkpoints/ssc-chat-v8-qwen3-4b/checkpoint-200
```

## Environment

- GPU: NVIDIA GB10 (sm_121, 128GB unified)
- Model: Qwen3-4B NF4 QLoRA (rank=16, alpha=32)
- Data: conversations_v4.jsonl (22,169 entries, 20.8% unsafe)
- Speed: 16 tok/s

## References

- Spec v12.22 (Run 8 started)
- entrenar#270 (RoPE+QK-norm fix — confirmed working, loss broke 6.66 floor)
- trueno#187 (NF4 GEMM speed — 16 tok/s bottleneck)

**Comment by noahgift** [comment]
Triage: P2 -- Training quality bug (STACK-006, 2026-03-31). Learning rate tuning for 22K dataset. May be mitigated by warmup cosine decay scheduler improvements. Needs empirical validation with current codebase.
